File: code-orig/stock-analysis-go/macdcalculator/macdcalculator.go
package macdcalculator

import (
	"stock-analysis-go/stockdata"
)

// CalculateMACDForDay calculates MACD for a specific day using the given stock data and MACD results array.
func CalculateMACDForDay(company []stockdata.StockData) float64 {

	ema12 := calculateEMA(company, 12)
	ema26 := calculateEMA(company, 26)

	return ema12 - ema26
}

// extractLast26Records extracts the last 26 records before the given day index for all companies.
func ExtractLast26Records(day int, company []stockdata.StockData) []stockdata.StockData {
	var extractedData []stockdata.StockData

	for i := day - 26; i < day; i++ {
		// Handle out-of-bounds index
		if i >= 0 {
			extractedData = append(extractedData, company[i])
		}
	}

	return extractedData
}

// calculateEMA calculates the Exponential Moving Average for the given stock data and period.
func calculateEMA(prices []stockdata.StockData, period int) float64 {
	if len(prices) < period {
		return 0.0
	}

	multiplier := 2.0 / float64(period+1)
	ema := prices[25].Price

	for i := 24 ; i >= 26 - period ; i-- {
		ema = (prices[i].Price-ema)*multiplier + ema
	}

	return ema
}

File: code-orig/stock-analysis-c/macd_calculator.c
#include "macd_calculator.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

StockData *
extract_last_26_records(StockData *source_data, int num_records, int start_index)
{
    int records_to_extract = 26;
    int extraction_start_index = (start_index - records_to_extract >= 0) ? (start_index - records_to_extract) : 0;

    StockData *extracted_data = malloc(records_to_extract * sizeof(StockData));
    if (extracted_data == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < records_to_extract; ++i)
    {
        extracted_data[i].date = source_data[extraction_start_index + i].date;
        extracted_data[i].price = source_data[extraction_start_index + i].price;
    }

    return extracted_data;
}

static double calculate_ema(StockData *prices, int period)
{
    double multiplier = 2.0 / (period + 1);
    double ema = prices[25].price; // Initial EMA is the closing price of the last day

    for (int i = 24; i >= 26 - period; --i)
    {
        ema = (prices[i].price - ema) * multiplier + ema;
    }

    return ema;
}

double calculate_macd_for_day(StockData *data)
{
    // Calculate 12-period and 26-period EMAs
    double ema12 = calculate_ema(data, 12);
    double ema26 = calculate_ema(data, 26);

    // Calculate MACD for the specific day
    double macd = ema12 - ema26;

    // Print the calculated MACD for the day
    // printf("MACD for the day: %.2f\n", macd);
    return macd;
}
